{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d881d517",
   "metadata": {},
   "source": [
    "# Tutorial for module discovery and characterization with CellSP\n",
    "\n",
    " - This notebook gives a basic understanding of CellSP's functionalities. There are 4 main steps involved:\n",
    "    - Data Preprocessing\n",
    "    - Subcellular Pattern Identification\n",
    "    - Module Discovery\n",
    "    - Module Characterization and Visualization\n",
    "\n",
    " - Please use Github Issues for any errors or feature requests.\n",
    "\n",
    "---\n",
    "\n",
    "### Installation\n",
    "\n",
    " - CellSP is avaliable on `pypi` and can be installed using `pip install cellSP`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "6a64945f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import cellSP"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60d75bfe",
   "metadata": {},
   "source": [
    "## Data Preprocessing\n",
    "\n",
    "CellSP analyzes single-molecule (subcellular) spatial transcriptomics (ST) data. To characterize module cells, we require scRNA-seq data from the same tissue to impute gene expression. Both datasets should be provided in the `AnnData` format.\n",
    "\n",
    "CellSP requires minimal preprocessing of input data. No preprocessing is performed on the input ST data; however, we recommend users apply quality control filtering of transcripts by appropriately thresholding Phred-scaled Q-Scores.\n",
    "\n",
    "For scRNA-seq data, we apply data denoising using `MAGIC` and perform imputation of ST data using `Tangram`.\n",
    "\n",
    "#### Loading `AnnData`\n",
    "\n",
    "CellSP provides the function `cellSP.ds.load_data()` to load `AnnData` objects. It takes the following arguments:\n",
    "\n",
    "- `sc_adata`: The `AnnData` object containing scRNA-seq data.\n",
    "- `st_adata`: The `AnnData` object containing ST data.\n",
    "\n",
    "These datasets are used for downstream analysis within CellSP.\n",
    "\n",
    "Example Datasets - An example data can be Xenium Whole mouse brain for [ST](https://www.10xgenomics.com/datasets/fresh-frozen-mouse-brain-replicates-1-standard) and Stero-seq for [SC](https://db.cngb.org/search/project/CNP0001543/). \n",
    "\n",
    "For this analysis, we use a subset of 5000 randomly sampled cells from the replicate 1 of the Xenium Whole Mouse Brain dataset available publically [link](https://www.10xgenomics.com/datasets/fresh-frozen-mouse-brain-replicates-1-standard). \n",
    "\n",
    "\n",
    "### File format\n",
    "\n",
    "The `sc_adata` file should follow the standard `AnnData` format for single-cell data, where:\n",
    "\n",
    "- `sc_adata.X` contains the gene expression matrix (cells Ã— genes).\n",
    "- `sc_adata.var` contains gene metadata, indexed by gene names.\n",
    "- `sc_adata.obs` contains cell metadata.\n",
    "\n",
    "The `st_adata` should have the same format as `sc_adata` and additionally is expected to have a DataFrame as the `transcripts` key in the `uns` attribute with the following columns - `['gene', 'uID', 'absX', 'absY']`. Also, the `obsm` attribute is expected to have another DataFrame which stores the spatial locations of the cells. \n",
    "\n",
    "Below is an example table for a 2D data (if the data is 3D, `absZ` column is also expected) - \n",
    "\n",
    "| gene | uID | absX | absY |\n",
    "| :---         |     :---:      |          ---: |           ---: |\n",
    "| AKAP11   |  2    | -1401.666    | -2956.618     |\n",
    "| SIPA1L3  |  3       | -1411.692      |   -2936.609     |\n",
    "| THBS1  |  925       | -764.6989      |   -1604.828    |\n",
    "\n",
    "Optionally, the user can also provide the cell boundaries of each cell. `cell_boundary` key in the `uns` key is expected to have a DataFrame with format as below - \n",
    "\n",
    "|uID|vertex_x|vertex_y|\n",
    "| :---         |          ---: |           ---: |\n",
    "|1|1554.65|2519.1875|\n",
    "|1|1550.825|2521.525|\n",
    "|2|1549.55|2523.4375|\n",
    "\n",
    "If cell boundaries are not provided, the convex hull of the transcripts of each cell are used as its approximation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f645efe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_sc, adata_st = cellSP.ds.load_data(sc_adata= 'adata_sc.h5ad', \n",
    "                                         st_adata = f\"adata_st.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9be1834a",
   "metadata": {},
   "source": [
    "For denoising of scRNA-seq data, we use the function `cellSP.pp.impute()` which has the following arguments,\n",
    " - `adata_sc`:  The `AnnData` object containing scRNA-seq data.\n",
    " - `t`: `MAGIC` parameter which corresponds to the power to which the diffusion operator is powered.\n",
    " - `**kwargs`: Additional arguments for `MAGIC`.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5a208f56",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculating MAGIC...\n",
      "  Running MAGIC on 50140 cells and 16168 genes.\n",
      "  Calculating graph and diffusion operator...\n",
      "    Calculating PCA...\n",
      "    Calculated PCA in 71.02 seconds.\n",
      "    Calculating KNN search...\n",
      "    Calculated KNN search in 64.78 seconds.\n",
      "    Calculating affinities...\n",
      "    Calculated affinities in 68.57 seconds.\n",
      "  Calculated graph and diffusion operator in 204.76 seconds.\n",
      "  Running MAGIC with `solver='exact'` on 16168-dimensional data may take a long time. Consider denoising specific genes with `genes=<list-like>` or using `solver='approximate'`.\n",
      "  Calculating imputation...\n",
      "    Automatically selected t = 4\n",
      "  Calculated imputation in 40.70 seconds.\n",
      "Calculated MAGIC in 246.68 seconds.\n",
      "Time to impute 0:04:06.696124\n"
     ]
    }
   ],
   "source": [
    "adata_sc = cellSP.pp.impute(adata_sc, t=\"auto\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b04c8fdc",
   "metadata": {},
   "source": [
    "For imputation of genes not in the ST data, we use the function `cellSP.pp.run_tangram()` which has the following arguments,\n",
    " - `adata_sc`:  The `AnnData` object containing scRNA-seq data.\n",
    " - `adata_st`:  The `AnnData` object containing ST data.\n",
    " - `device`: Device to run Tangram on. Either 'cpu' or 'cuda'."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "1b933f64",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Tangram...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:16168 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:16168 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.\n",
      "INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.\n",
      "INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.\n",
      "INFO:root:Allocate tensors for mapping.\n",
      "INFO:root:Begin training with 16168 genes and rna_count_based density_prior in cells mode...\n",
      "INFO:root:Printing scores every 100 epochs.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score: 0.780, KL reg: 0.219\n",
      "Score: 0.987, KL reg: 0.002\n",
      "Score: 0.994, KL reg: 0.001\n",
      "Score: 0.995, KL reg: 0.001\n",
      "Score: 0.996, KL reg: 0.001\n",
      "Score: 0.997, KL reg: 0.000\n",
      "Score: 0.997, KL reg: 0.000\n",
      "Score: 0.997, KL reg: 0.000\n",
      "Score: 0.997, KL reg: 0.000\n",
      "Score: 0.998, KL reg: 0.000\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Saving results..\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time to run Tangram 0:26:53.856380\n"
     ]
    }
   ],
   "source": [
    "adata_st = cellSP.pp.run_tangram(adata_sc, adata_st, device='cuda')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66bd4aef",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "### Saving the `AnnData` object.\n",
    "\n",
    "CellSP provides the function `cellSP.ds.write_h5ad()` to save `AnnData` objects. It takes the following arguments:\n",
    "- `data`: Anndata object to save.\n",
    "- `filename`: Name of the file to save."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "fe6d8f60",
   "metadata": {},
   "outputs": [],
   "source": [
    "cellSP.io.write_h5ad(adata_st, f\"adata_st.h5ad\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ceaa62d7",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Subcellular Pattern Identification\n",
    "\n",
    "CellSP utilizes existing tools [InSTAnT](https://github.com/bhavaygg/InSTAnT/) and [SPRAWL](https://github.com/salzman-lab/SPRAWL/) for the discovery of subcellular spatial patterns of genes within each cell. Types of patterns detected include gene-gene colocalization reported by InSTAnT and four types of subcellular localization preferences (peripheral, punctate, central, radial) reported by SPRAWL.\n",
    "\n",
    "#### Running `InSTAnT`.\n",
    "\n",
    "We use the function `cellSP.ch.run_instant()`, which has the following parameters:\n",
    "- **`adata_st`** (`AnnData`): The `AnnData` object containing spatial transcriptomics data.  \n",
    "- **`distance_threshold`** (`float`): The distance threshold for the PP-Test.  \n",
    "- **`threads`** (`int`): The number of threads to use.  \n",
    "- **`n_vertices`** (`int`, optional): The number of vertices to use for the FSM (default: `None`).  \n",
    "- **`alpha_cpb`** (`float`, optional): The significance level for the CPB (default: `0.001`).  \n",
    "- **`filename`** (`str`, optional): The name of the file to save the preprocessed data (default: `.cellSP_st.h5ad`, deleted after completion).  \n",
    "- **`remove_file`** (`bool`, optional): If `True`, the file is deleted after completion (default: `True`).  \n",
    "- **`is_sliced`** (`bool`, optional): Specifies whether the data is sliced (default: `True`).  \n",
    "\n",
    "We create a temporary `Anndata` file to run `InSTAnT` with name `filename`. If the argument is not specified, we use a random name and delete the file unless specified otherwise in `remove_file`. The argument `is_sliced` is used to denote whether the ST data has discrete z-planes (like in some MERFISH experiments) or continous (like in Xenium). \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6c8734dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running InSTAnT...\n",
      "Loaded Data. Number of Transcripts:  1909167 , Number of Genes:  248 , Number of Cells:  5000\n",
      "Initialised PP-3D slice now on 128 threads\n",
      "Number of cells:  5000 , Number of Genes:  248\n",
      "min genecount less than 20 for cell id 622, Skipping ...\n",
      "min genecount less than 20 for cell id 623, Skipping ...\n",
      "min genecount less than 20 for cell id 716, Skipping ...\n",
      "min genecount less than 20 for cell id 1253, Skipping ...\n",
      "min genecount less than 20 for cell id 3878, Skipping ...\n",
      "min genecount less than 20 for cell id 4858, Skipping ...\n",
      "Running PP-3D slice now on 128 threads for, 4994 cells, 1909167 transcripts\n",
      "Cell-wise Proximal Pairs Time : 3618.63 seconds\n",
      "Running Global Colocalization now on 128 threads\n",
      "Number of cells: 5000, Number of genes: 248\n",
      "Global Colocalization initialized ..\n",
      "High Precision Global Colocalization Time: 59.71 seconds\n",
      "Cell-wise Global Colocalization Time : 122.07 seconds\n"
     ]
    }
   ],
   "source": [
    "adata_st = cellSP.ch.run_instant(adata_st = adata_st, distance_threshold=2, threads=128)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff4be272",
   "metadata": {},
   "source": [
    "#### Running `SPRAWL`.\n",
    "\n",
    "We use the function `cellSP.ch.run_sprawl()`, which has the following parameters:\n",
    "- **`adata_st`** (`AnnData`): The `AnnData` object containing spatial transcriptomics data.  \n",
    "- **`methods`** (`list`, optional): List of methods/patterns to run SPRAWL on. Options: `Peripheral`, `Radial`, `Punctate`, `Central` (default: `['Peripheral', 'Radial', 'Punctate', 'Central']`).  \n",
    "- **`threads`** (`int`, optional): Number of threads to use (default: `1`).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e9bcff87",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running SPRAWL...\n",
      "SPRAWL completed in: 0:52:01.080050\n"
     ]
    }
   ],
   "source": [
    "adata_st = cellSP.ch.run_sprawl(adata_st, threads = 128)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f175df04",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Module Discovery\n",
    "\n",
    "Patterns identified in the previous step (represented as matrices of cells Ã— genes or gene pairs) undergo biclustering analysis to identify subsets of rows and columns with high average values. These subsets represent **gene-cell modules**, where genes or gene pairs exhibit the same subcellular pattern within the same cells. Gene-cell modules that significantly overlap in genes (or gene pairs) are merged to reduce redundancy. This step outputs a set of gene-cell modules with varying numbers of cells and genes.\n",
    "\n",
    "CellSP uses the biclustering algorithm `LAS` to analyze each pattern and identify gene-cell modules.\n",
    "\n",
    "#### Module Discovery Functions\n",
    "\n",
    "CellSP provides two functions for module discovery, one for `SPRAWL` (`cellSP.ch.bicluster_sprawl()`) and one for `InSTAnT` (`cellSP.ch.bicluster_instant()`). \n",
    "\n",
    "Both functions share the same parameters, but the `InSTAnT` function includes two additional parameters:\n",
    "\n",
    "- **`alpha`** (`float`, optional): p-value significance threshold below which a gene pair is considered for biclustering (default: `1e-3`).  \n",
    "- **`topk`** (`int`, optional): Selects only the `K` most significant gene pairs with p-values < `alpha` (default: `None`).  \n",
    "\n",
    "These parameters help restrict the number of gene pairs for biclustering, reducing computational complexity.\n",
    "\n",
    "#### Shared Parameters\n",
    "\n",
    "- **`adata_st`** (`AnnData`): The `AnnData` object containing spatial transcriptomics data.  \n",
    "- **`num_biclusters`** (`int` or `auto`, optional): Number of modules to find (default: `auto`).  \n",
    "- **`randomized_searches`** (`int`, optional): Number of randomized searches to perform in LAS (default: `50,000`).  \n",
    "- **`gene_threshold`** (`int`, optional): Minimum number of genes required for a valid bicluster (default: `3`).  \n",
    "- **`threads`** (`int`, optional): Number of threads to use for parallel computation (default: `1`).  \n",
    "- **`expand`** (`bool`, optional): Whether to expand biclusters by including additional nearby or correlated entries (default: `True`).  \n",
    "- **`oc`** (`float`, optional): Overlap coefficient threshold for merging overlapping biclusters (default: `0.667`).\n",
    "\n",
    "\n",
    "**Note** - Using `auto` mode for bilcustering generates a null distribution for each pattern and finds suitable LAS score threshold based on the data. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "70777f10",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bi-clustering InSTAnT CPB results...\n",
      "InSTAnT CPB Bi-clustering time: 1254 days, 0:10:50.194914\n"
     ]
    }
   ],
   "source": [
    "adata_st = cellSP.ch.bicluster_instant(adata_st, distance_threshold=2, threads=128, alpha=1e-5, num_biclusters = 10, randomized_searches = 50000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9d2df827",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bi-clustering SPRAWL results...\n",
      "SPRAWL bi-clustering completed time: 1:12:12.287354\n"
     ]
    }
   ],
   "source": [
    "adata_st = cellSP.ch.bicluster_sprawl(adata_st, threads=128, num_biclusters = 10, randomized_searches = 50000)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "238ed8c4",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "## Module Characterization and Visualization\n",
    "\n",
    "### Module Characterization\n",
    "\n",
    "To aid biological interpretation, CellSP reports shared properties of the genes and cells of each discovered module. Genes are characterized using Gene Ontology (GO) enrichment tests, while cells are characterized by their cell type composition if such information is available. To provide a more precise characterization of a moduleâ€™s cells, CellSP trains a machine learning classifier to discriminate those cells from all other cells, using the expression levels of all genes other than the module genes. Genes that are highly predictive in this task are then subjected to GO enrichment tests, furnishing hypotheses about biological processes and pathways that are active specifically in the module cells.\n",
    "\n",
    "To characterize the module genes, we perform GO enrichment using `cellSP.geo.geo_analysis()` with the paramter `setting` set to `module`. The detailed arguments are mentioned below.\n",
    "\n",
    "- **`adata_st`** (`AnnData`): The `AnnData` object containing spatial transcriptomics data.  \n",
    "- **`mode`** (`list`, optional): List of analyses to perform. Must be one or more of `['instant_fsm', 'instant_biclustering', 'sprawl_biclustering']` (default: `['instant_biclustering', 'sprawl_biclustering']`).  \n",
    "- **`organism`** (`int`, optional): Taxon ID of the organism (default: `10090`).  \n",
    "- **`do_revigo`** (`bool`, optional): If `True`, performs REVIGO analysis (default: `True`).  \n",
    "- **`setting`** (`str`, optional): Specifies whether to perform the analysis at the **module** or **cell** level. Must be `'module'` or `'cell'` (default: `'module'`).  \n",
    "- **`corr_threshold`** (`float`, optional): Correlation threshold for finding genes correlated to the marker genes of the module cells (default: `0.98`).  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "110cd68c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Performing GO Enrichment Analysis...\n",
      "GO Enrichment Analysis Completed in : 0:10:23.404245\n"
     ]
    }
   ],
   "source": [
    "adata_st = cellSP.geo.geo_analysis(adata_st, setting=\"module\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36ad26b4",
   "metadata": {},
   "source": [
    "To characterize the module cells, we train a Random Forest classifier of module cells using the function `cellSP.md.model_modules()`. Its arguments are - \n",
    "\n",
    "- **`adata_st`** (`AnnData`): The `AnnData` object containing spatial transcriptomics data.  \n",
    "- **`mode`** (`list`, optional): List of characterizations to model. Options: `['instant_biclustering', 'sprawl_biclustering']` (default: `['instant_biclustering', 'sprawl_biclustering']`).  \n",
    "- **`n_repeats`** (`int`, optional): Number of sampling repetitions (default: `25`).  \n",
    "- **`subsample`** (`bool`, optional): If `True`, subsamples cells based on cell type (if available) or randomly (default: `True`).  \n",
    "- **`corr_threshold`** (`float`, optional): Correlation threshold for filtering gene sets (default: `0.98`).\n",
    "\n",
    "The marker genes for each module are found using `SHAP` and are used to provide biological interpretation of the module cells beyond their cell type."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e6a55b46",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Modeling subcellular patterns...\n",
      "Modelling for Module 0 with genes: ['cldn5', 'ly6a', 'paqr5', 'fgd5', 'acvrl1', 'pecam1', 'kdr', 'cobll1', 'pglyrp1', 'car4', 'slfn5', 'sox17', 'nostrin', 'mecom', 'zfp366', 'ccn2', 'adgrl4', 'emcn', 'fn1', 'cd93'] in 423 cells\n",
      "Modelling for Module 1 with genes: ['lyz2', 'spi1', 'cd300c2', 'laptm5', 'trem2', 'plekha2', 'arhgap25', 'cd53', 'siglech', 'kctd12', 'ikzf1', 'cd68'] in 211 cells\n",
      "Modelling for Module 2 with genes: ['col1a1', 'igf2', 'lyz2', 'cyp1b1', 'spp1', 'pdgfra', 'ccn2', 'gjb2', 'dcn', 'col6a1', 'fn1', 'fmod', 'slc13a4', 'igfbp4', 'aldh1a2'] in 127 cells\n",
      "Modelling for Module 3 with genes: ['opalin', 'sox10', 'tmem163', 'zfp536', 'clmn', 'dpy19l1', 'sema6a', 'prox1', 'adamtsl1', 'gng12', 'gjc3'] in 303 cells\n",
      "Modelling for Module 4 with genes: ['sema3d', 'sox10', 'cspg4', 'pdgfra', 'gpr17', 'pou3f1', 'tmem255a', 'sema5b', 'gng12', 'gjc3'] in 120 cells\n",
      "Modelling for Module 5 with genes: ['cbln4', 'thsd7a', 'gad1', 'kcnmb2', 'sst', 'sema3e', 'necab1', 'gad2', 'rbp4', 'btbd11', 'parm1', 'rab3b', 'pvalb', 'plcxd3', 'cdh13', 'dner', 'vat1l', 'pdyn', 'cort'] in 134 cells\n",
      "Modelling for Module 6 with genes: ['slc39a12', 'ntsr2', 'gli3', 'rmst', 'aqp4', 'clmn', 'acsbg1', 'hapln1', 'id2', 'rfx4'] in 286 cells\n",
      "Modelling for Module 7 with genes: ['acta2', 'cspg4', 'ccn2', 'carmn', 'inpp4b', 'ano1', 'car4'] in 34 cells\n",
      "Modelling for Module 0 with genes: ['slc13a4', 'aldh1a2', 'spp1', 'fmod', 'dcn', 'igf2', 'acta2'] in 61 cells\n",
      "Modelling for Module 1 with genes: ['cldn5', 'adgrl4', 'pecam1', 'acvrl1', 'ly6a', 'fn1', 'pglyrp1', 'sox17', 'kdr', 'emcn', 'nostrin', 'cd93', 'zfp366'] in 181 cells\n",
      "Modelling for Module 2 with genes: ['cd53', 'trem2', 'cd300c2', 'siglech', 'cd68', 'ikzf1', 'spi1', 'laptm5'] in 89 cells\n",
      "Modelling for Module 3 with genes: ['aldh1a2', 'col1a1', 'fmod', 'spp1', 'gjb2', 'cyp1b1', 'slc13a4', 'dcn', 'igf2'] in 79 cells\n",
      "Modelling for Module 4 with genes: ['ano1', 'carmn', 'cspg4'] in 57 cells\n",
      "Modelling for Module 5 with genes: ['gpr17', 'cspg4', 'sema3d', 'pdgfra'] in 104 cells\n",
      "Modelling for Module 6 with genes: ['cldn5', 'adgrl4', 'pecam1', 'acvrl1', 'ly6a', 'pglyrp1', 'fn1', 'kdr', 'sox17', 'emcn', 'cd93', 'nostrin', 'mecom', 'fgd5'] in 204 cells\n",
      "Modelling for Module 7 with genes: ['cd300c2', 'cd53', 'trem2', 'ikzf1', 'siglech', 'spi1', 'cd68', 'laptm5'] in 99 cells\n",
      "Modelling for Module 8 with genes: ['cspg4', 'gpr17', 'ano1', 'carmn', 'sema3d', 'pdgfra'] in 193 cells\n",
      "Modelling for Module 9 with genes: ['fmod', 'col1a1', 'spp1', 'aldh1a2', 'dcn', 'gjb2', 'cyp1b1', 'igf2', 'slc13a4', 'acta2'] in 88 cells\n",
      "Modelling for Module 10 with genes: ['sox10', 'opalin', 'gjc3', 'zfp536', 'sema6a'] in 307 cells\n",
      "Modelling for Module 11 with genes: ['cort', 'sst', 'pthlh', 'kcnmb2', 'rbp4', 'gad2', 'crh', 'rab3b', 'chodl', 'gad1', 'trpc4', 'eya4', 'pdyn', 'col19a1', 'igf1', 'cntnap4', 'pvalb', 'thsd7a'] in 126 cells\n",
      "Modelling for Module 12 with genes: ['cldn5', 'pglyrp1', 'ly6a', 'pecam1', 'acvrl1', 'adgrl4', 'emcn', 'fn1', 'sox17', 'cd93', 'kdr', 'fgd5', 'mecom', 'zfp366', 'nostrin'] in 150 cells\n",
      "Modelling for Module 13 with genes: ['trem2', 'cd300c2', 'cd53', 'siglech', 'ikzf1', 'cd68', 'spi1', 'laptm5'] in 93 cells\n",
      "Modelling for Module 14 with genes: ['cspg4', 'gpr17', 'carmn', 'ano1', 'sema3d', 'pdgfra'] in 136 cells\n",
      "Modelling for Module 15 with genes: ['opalin', 'sox10', 'gjc3', 'zfp536', 'sema6a', 'dpy19l1', 'prox1'] in 193 cells\n",
      "Time to model subcellular patterns 0:45:14.681274\n"
     ]
    }
   ],
   "source": [
    "adata_st = cellSP.md.model_modules(adata_st, do_shap=True, subsample = True, corr_threshold = 0.90, n_repeats = 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7810a2bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Performing GO Enrichment Analysis...\n",
      "GO Enrichment Analysis Completed in : 0:31:44.734452\n"
     ]
    }
   ],
   "source": [
    "adata_st = cellSP.geo.geo_analysis(adata_st, setting=\"cell\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "facf5c7b",
   "metadata": {},
   "source": [
    "### Module Visualization\n",
    "\n",
    "CellSP provides functionality to visualize the detected patterns for each module and also to visualize each pattern on each module.\n",
    "\n",
    "The first function, `cellSP.vs.visualize_modules()`, visualizes all modules for their detected patterns. Its arguments are - \n",
    "\n",
    "- **adata_st** (*AnnData*): Spatial transcriptomics data.\n",
    "- **mode** (*list*): List of characterizations. (`'instant_biclustering'`, `'sprawl_biclustering'`).\n",
    "- **num_sectors** (*int*): Number of sectors in the circular cell (default: `10`).\n",
    "- **num_ccircles** (*int*): Number of  Concentric circles (default: `5`).\n",
    "- **distance_threshold** (*float*): PP-Test distance. Used to create proximity heatmap. (default: `2`).\n",
    "- **positions** (*bool*): Plot spatial positions of module cells in a subplot. (`True`).\n",
    "- **is_sliced** (*bool*): Are the z-slices discrete? (`True`).\n",
    "\n",
    "The next function, `cellSP.vs.visualize_pattern()` allows you to visualize any pattern and save it for a given module. Its arguments are - \n",
    "\n",
    "- **adata_st** (*AnnData*): Spatial transcriptomics data.  \n",
    "- **module_number** (*int*): Index of the module to visualize.  \n",
    "- **pattern** (*str*): Spatial pattern (`'Proximal'`, `'Radial'`, `'Concentric'`).  \n",
    "- **mode** (*list*): List of characterizations. (`'instant_biclustering'`, `'sprawl_biclustering'`).\n",
    "- **filename** (*str*): Filename to save the plot.  (`None`)\n",
    "- **num_sectors** (*int*): Number of sectors in the circular cell (default: `10`).\n",
    "- **num_ccircles** (*int*): Number of  Concentric circles (default: `5`).\n",
    "- **distance_threshold** (*float*): PP-Test distance. Used to create proximity heatmap. (default: `2`).\n",
    "\n",
    "CellSP creates a comprehensive webpage summary for all detected modules. The report showcases key details such as the identified subcellular pattern, spatial distribution of module cells in the tissue, UMAP visualization of the module cells, and the functional characterization of the module. We use the function `cellSP.vs.create_report()` which require the `adata_st` as input and runs the pattern visualization and other plotting functions to generate a HTML report. The report generated for this example is available as PDF [here](https://github.com/bhavaygg/CellSP/blob/main/figures/example_report.pdf). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66709af0",
   "metadata": {},
   "outputs": [],
   "source": [
    "cellSP.vs.create_report(adata_st)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f57fda4e",
   "metadata": {},
   "source": [
    "--- \n",
    "\n",
    "## `AnnData` format\n",
    "\n",
    "The `AnnData` file is updated with the outputs of each of the functions. Here we try to explain what each output holds and means"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07aa753b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs Ã— n_vars = 5000 Ã— 16171\n",
       "    obs: 'region', 'uniform_density', 'rna_count_based_density'\n",
       "    uns: 'cell_boundary', 'cpb_results', 'geneList', 'instant_biclustering', 'instant_biclustering_geo_cell', 'instant_biclustering_geo_module', 'instant_fsm', 'instant_fsm_geo_cell', 'instant_fsm_geo_module', 'nV4_cliques', 'overlap_genes', 'pp_test_d2_pvalues', 'sprawl_biclustering', 'sprawl_biclustering_geo_cell', 'sprawl_biclustering_geo_module', 'sprawl_scores', 'training_genes', 'transcripts'\n",
       "    obsm: 'genecount', 'spatial'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "adata_st"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3485549",
   "metadata": {},
   "source": [
    "- `obs`\n",
    "    - `region`: ID of the region where the cell lies. (Not used)\n",
    "    - `cell_type`: Cell type of the cell. Used for sampling during cell characterization.\n",
    "    - `uniform_density` and `rna_count_based_density`: Tangram outputs.\n",
    "- `uns`\n",
    "    - `transcripts`: `DataFrame` with columns `uID` (cell ID), `absX`, `absY`, `absZ`(Optional), 'gene'.\n",
    "    - `cpb_results`: InSTAnT outputs. CPB Test results (DataFrame), `geneList`: genes used (list), `nV4_cliques`: FSM output (cliques), `pp_test_d2_pvalues`: PP Test p-values (NumPy matrix).\n",
    "    - `sprawl_scores`: SPRAWL output. Dictionary with the patterns as keys for which SPRAWL was run for. Values are numpy matrices of shape n_cells X n_genes. \n",
    "    - `cell_boundary`: `DataFrame` with columns `vertex_x` and `vertex_y`. Used in `SPRAWL` and plotting. If not present, a convex hull of the transcripts is used as cell boundary. \n",
    "    - `training_genes` and `overlap_genes`: Tangram outputs.\n",
    "    - `instant_biclustering` and `sprawl_biclustering`: Outputs of module discovery step. Described below.\n",
    "    - `instant_biclustering_geo_{}` and `sprawl_biclustering_geo_{}`: Outputs of module characterization step. Described below.\n",
    "- `obsm`:\n",
    "    - `spatial`: `DataFrame` with columns `absX`, `absY`, `absZ`(Optional) with the spatial location of each cell in the tissue.\n",
    "    - `X-UMAP`: `DataFrame` with columns `UMAP-1`, `UMAP-2` corresponding to the first and second UMAP dimension of each cell. Used for plotting in report."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f517a9aa",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "#### Module Discovery Outputs\n",
    "\n",
    "Outputs from biclustering on `InSTAnT` and `SPRAWL` are presented as `DataFrames` as shown below. Each row represents a detected module. The columns denote\n",
    "- `gene-pairs`: (for InSTAnT) Pairs of genes in the module.\n",
    "- `genes`: Genes in the module\n",
    "- `uIDs`: Cells in the module\n",
    "- `#cells`: Number of cells in the module.\n",
    "- `combined`: Number of times the module was merged.\n",
    "- `tangram`: Average cross-validation accuracy for module classification using genes other than module genes.\n",
    "- `tangram_corr`: Average cross-validation accuracy for module classification using genes other than module genes and genes correlated with them.\n",
    "- `baseline`: Average cross-validation accuracy for module classification using module genes.\n",
    "- `shap_genes`: Top 20 most predictive genes calculated using `SHAP`, also called as marker genes of the module cells.\n",
    "- `GO Module`: Most significant GO term found using enrichment test of module genes.\n",
    "- `GO Cell`: Most significant GO term found using enrichment test of marker genes of module cells.\n",
    "- `#pc_genes`: Number of genes found to be highly correlated with the marker genes of the module cells and are used for characterization of the module cells.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "6b8f8def",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene-pairs</th>\n",
       "      <th>genes</th>\n",
       "      <th>uIDs</th>\n",
       "      <th>#cells</th>\n",
       "      <th>combined</th>\n",
       "      <th>pre cell expansion</th>\n",
       "      <th>post cell expansion</th>\n",
       "      <th>instant average</th>\n",
       "      <th>instant score</th>\n",
       "      <th>tangram</th>\n",
       "      <th>tangram_corr</th>\n",
       "      <th>baseline</th>\n",
       "      <th>shap genes</th>\n",
       "      <th>GO Module</th>\n",
       "      <th>#pc_genes</th>\n",
       "      <th>GO Cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>(fgd5,pecam1),(fgd5,nostrin),(nostrin,zfp366),...</td>\n",
       "      <td>cldn5,ly6a,paqr5,fgd5,acvrl1,pecam1,kdr,cobll1...</td>\n",
       "      <td>140436,129048,5309,39156,141970,45390,140846,9...</td>\n",
       "      <td>423</td>\n",
       "      <td>1</td>\n",
       "      <td>34133.725391</td>\n",
       "      <td>34133.725391</td>\n",
       "      <td>0.8341564079291461,0.7305973067255646:0.637820...</td>\n",
       "      <td>34133.725390515654,34133.725390515654:31150.94...</td>\n",
       "      <td>0.879667</td>\n",
       "      <td>0.876585</td>\n",
       "      <td>0.873298</td>\n",
       "      <td>col1a2,srgn,pik3cg,myoc,fmod,acta2,cp,mpzl2,fo...</td>\n",
       "      <td>angiogenesis</td>\n",
       "      <td>59.0</td>\n",
       "      <td>collagen fibril organization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>(plekha2,trem2),(laptm5,trem2),(cd68,spi1),(kc...</td>\n",
       "      <td>lyz2,spi1,cd300c2,laptm5,trem2,plekha2,arhgap2...</td>\n",
       "      <td>79722,95083,13368,149578,13804,156896,100660,1...</td>\n",
       "      <td>211</td>\n",
       "      <td>1</td>\n",
       "      <td>24778.336737</td>\n",
       "      <td>24778.336737</td>\n",
       "      <td>1.3074074013208965,0.9338975380662246:0.884153...</td>\n",
       "      <td>24778.33673746181,24778.33673746181:21730.8592...</td>\n",
       "      <td>0.839557</td>\n",
       "      <td>0.847780</td>\n",
       "      <td>0.834535</td>\n",
       "      <td>gm5136,gm19426,bcl2a1b,dgkq,itgb2,gm2582,17001...</td>\n",
       "      <td>cellular response to lipopolysaccharide</td>\n",
       "      <td>32.0</td>\n",
       "      <td>extracellular matrix organization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>(col1a1,igf2), (aldh1a2,igf2), (aldh1a2,fmod),...</td>\n",
       "      <td>col1a1,igf2,lyz2,cyp1b1,spp1,pdgfra,ccn2,gjb2,...</td>\n",
       "      <td>40682,8215,3804,2392,155150,30568,10615,14188,...</td>\n",
       "      <td>127</td>\n",
       "      <td>0</td>\n",
       "      <td>16971.652816</td>\n",
       "      <td>16971.652816</td>\n",
       "      <td>0.7376447113799066</td>\n",
       "      <td>16971.652815512436</td>\n",
       "      <td>0.944877</td>\n",
       "      <td>0.932662</td>\n",
       "      <td>0.942615</td>\n",
       "      <td>nfatc4,slc22a6,olfr1033,tfpi,adam12,stra6,aebp...</td>\n",
       "      <td>response to hormone</td>\n",
       "      <td>61.0</td>\n",
       "      <td>skeletal system development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>(opalin,sox10), (gjc3,opalin), (clmn,opalin), ...</td>\n",
       "      <td>opalin,sox10,tmem163,zfp536,clmn,dpy19l1,sema6...</td>\n",
       "      <td>115752,30,103084,112906,48454,73187,128197,113...</td>\n",
       "      <td>303</td>\n",
       "      <td>0</td>\n",
       "      <td>15972.587709</td>\n",
       "      <td>15972.587709</td>\n",
       "      <td>0.9100920497163525</td>\n",
       "      <td>15972.587709481668</td>\n",
       "      <td>0.918033</td>\n",
       "      <td>0.909601</td>\n",
       "      <td>0.913104</td>\n",
       "      <td>gm42756,d7ertd443e,smco3,gm44866,sec14l5,gm459...</td>\n",
       "      <td>ensheathment of neurons</td>\n",
       "      <td>76.0</td>\n",
       "      <td>myelination</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>(cspg4,pdgfra), (cspg4,gpr17), (pdgfra,sema3d)...</td>\n",
       "      <td>sema3d,sox10,cspg4,pdgfra,gpr17,pou3f1,tmem255...</td>\n",
       "      <td>46594,128197,88969,120518,20507,94389,1377,906...</td>\n",
       "      <td>120</td>\n",
       "      <td>0</td>\n",
       "      <td>13568.407786</td>\n",
       "      <td>13568.407786</td>\n",
       "      <td>1.218629438706812</td>\n",
       "      <td>13568.407785650978</td>\n",
       "      <td>0.827500</td>\n",
       "      <td>0.812500</td>\n",
       "      <td>0.821667</td>\n",
       "      <td>ermn,creb5,c030029h02rik,tsc22d4,cpm,d7ertd443...</td>\n",
       "      <td>ensheathment of neurons</td>\n",
       "      <td>92.0</td>\n",
       "      <td>ensheathment of neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>(gad1,rab3b), (gad1,gad2), (gad2,rab3b), (dner...</td>\n",
       "      <td>cbln4,thsd7a,gad1,kcnmb2,sst,sema3e,necab1,gad...</td>\n",
       "      <td>19699,159779,61459,62491,143327,15746,96168,40...</td>\n",
       "      <td>134</td>\n",
       "      <td>0</td>\n",
       "      <td>9092.400214</td>\n",
       "      <td>9092.400214</td>\n",
       "      <td>0.9149824103740001</td>\n",
       "      <td>9092.400214200303</td>\n",
       "      <td>0.776439</td>\n",
       "      <td>0.786111</td>\n",
       "      <td>0.786952</td>\n",
       "      <td>prokr2,npy,6530411m01rik,npas1,catip,slc32a1,g...</td>\n",
       "      <td>glutamate catabolic process</td>\n",
       "      <td>31.0</td>\n",
       "      <td>cerebral cortex GABAergic interneuron differen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>(aqp4,ntsr2), (aqp4,slc39a12), (ntsr2,slc39a12...</td>\n",
       "      <td>slc39a12,ntsr2,gli3,rmst,aqp4,clmn,acsbg1,hapl...</td>\n",
       "      <td>45322,112631,69762,79737,46629,107264,120534,5...</td>\n",
       "      <td>286</td>\n",
       "      <td>0</td>\n",
       "      <td>5291.380998</td>\n",
       "      <td>5291.380998</td>\n",
       "      <td>0.7901200793394628</td>\n",
       "      <td>5291.380998292888</td>\n",
       "      <td>0.825188</td>\n",
       "      <td>0.823436</td>\n",
       "      <td>0.821310</td>\n",
       "      <td>ac034116.3,phka1,ppara,nat8f3,ccdc141,gm35552,...</td>\n",
       "      <td>regulation of smoothened signaling pathway</td>\n",
       "      <td>27.0</td>\n",
       "      <td>regulation of generation of precursor metaboli...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>(ano1,carmn), (car4,carmn), (cspg4,inpp4b), (c...</td>\n",
       "      <td>acta2,cspg4,ccn2,carmn,inpp4b,ano1,car4</td>\n",
       "      <td>78280,26700,127795,107900,10612,80885,61426,14...</td>\n",
       "      <td>34</td>\n",
       "      <td>0</td>\n",
       "      <td>4333.368398</td>\n",
       "      <td>4333.368398</td>\n",
       "      <td>1.110971154375574</td>\n",
       "      <td>4333.368397747355</td>\n",
       "      <td>0.848333</td>\n",
       "      <td>0.853571</td>\n",
       "      <td>0.955714</td>\n",
       "      <td>crip1,cp,cavin1,myh11,cfh,pf4,gm47719,mpzl2,s1...</td>\n",
       "      <td>positive regulation of MAPK cascade</td>\n",
       "      <td>79.0</td>\n",
       "      <td>tissue development</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          gene-pairs  \\\n",
       "0  (fgd5,pecam1),(fgd5,nostrin),(nostrin,zfp366),...   \n",
       "1  (plekha2,trem2),(laptm5,trem2),(cd68,spi1),(kc...   \n",
       "2  (col1a1,igf2), (aldh1a2,igf2), (aldh1a2,fmod),...   \n",
       "3  (opalin,sox10), (gjc3,opalin), (clmn,opalin), ...   \n",
       "4  (cspg4,pdgfra), (cspg4,gpr17), (pdgfra,sema3d)...   \n",
       "5  (gad1,rab3b), (gad1,gad2), (gad2,rab3b), (dner...   \n",
       "6  (aqp4,ntsr2), (aqp4,slc39a12), (ntsr2,slc39a12...   \n",
       "7  (ano1,carmn), (car4,carmn), (cspg4,inpp4b), (c...   \n",
       "\n",
       "                                               genes  \\\n",
       "0  cldn5,ly6a,paqr5,fgd5,acvrl1,pecam1,kdr,cobll1...   \n",
       "1  lyz2,spi1,cd300c2,laptm5,trem2,plekha2,arhgap2...   \n",
       "2  col1a1,igf2,lyz2,cyp1b1,spp1,pdgfra,ccn2,gjb2,...   \n",
       "3  opalin,sox10,tmem163,zfp536,clmn,dpy19l1,sema6...   \n",
       "4  sema3d,sox10,cspg4,pdgfra,gpr17,pou3f1,tmem255...   \n",
       "5  cbln4,thsd7a,gad1,kcnmb2,sst,sema3e,necab1,gad...   \n",
       "6  slc39a12,ntsr2,gli3,rmst,aqp4,clmn,acsbg1,hapl...   \n",
       "7            acta2,cspg4,ccn2,carmn,inpp4b,ano1,car4   \n",
       "\n",
       "                                                uIDs  #cells  combined  \\\n",
       "0  140436,129048,5309,39156,141970,45390,140846,9...     423         1   \n",
       "1  79722,95083,13368,149578,13804,156896,100660,1...     211         1   \n",
       "2  40682,8215,3804,2392,155150,30568,10615,14188,...     127         0   \n",
       "3  115752,30,103084,112906,48454,73187,128197,113...     303         0   \n",
       "4  46594,128197,88969,120518,20507,94389,1377,906...     120         0   \n",
       "5  19699,159779,61459,62491,143327,15746,96168,40...     134         0   \n",
       "6  45322,112631,69762,79737,46629,107264,120534,5...     286         0   \n",
       "7  78280,26700,127795,107900,10612,80885,61426,14...      34         0   \n",
       "\n",
       "   pre cell expansion  post cell expansion  \\\n",
       "0        34133.725391         34133.725391   \n",
       "1        24778.336737         24778.336737   \n",
       "2        16971.652816         16971.652816   \n",
       "3        15972.587709         15972.587709   \n",
       "4        13568.407786         13568.407786   \n",
       "5         9092.400214          9092.400214   \n",
       "6         5291.380998          5291.380998   \n",
       "7         4333.368398          4333.368398   \n",
       "\n",
       "                                     instant average  \\\n",
       "0  0.8341564079291461,0.7305973067255646:0.637820...   \n",
       "1  1.3074074013208965,0.9338975380662246:0.884153...   \n",
       "2                                 0.7376447113799066   \n",
       "3                                 0.9100920497163525   \n",
       "4                                  1.218629438706812   \n",
       "5                                 0.9149824103740001   \n",
       "6                                 0.7901200793394628   \n",
       "7                                  1.110971154375574   \n",
       "\n",
       "                                       instant score   tangram  tangram_corr  \\\n",
       "0  34133.725390515654,34133.725390515654:31150.94...  0.879667      0.876585   \n",
       "1  24778.33673746181,24778.33673746181:21730.8592...  0.839557      0.847780   \n",
       "2                                 16971.652815512436  0.944877      0.932662   \n",
       "3                                 15972.587709481668  0.918033      0.909601   \n",
       "4                                 13568.407785650978  0.827500      0.812500   \n",
       "5                                  9092.400214200303  0.776439      0.786111   \n",
       "6                                  5291.380998292888  0.825188      0.823436   \n",
       "7                                  4333.368397747355  0.848333      0.853571   \n",
       "\n",
       "   baseline                                         shap genes  \\\n",
       "0  0.873298  col1a2,srgn,pik3cg,myoc,fmod,acta2,cp,mpzl2,fo...   \n",
       "1  0.834535  gm5136,gm19426,bcl2a1b,dgkq,itgb2,gm2582,17001...   \n",
       "2  0.942615  nfatc4,slc22a6,olfr1033,tfpi,adam12,stra6,aebp...   \n",
       "3  0.913104  gm42756,d7ertd443e,smco3,gm44866,sec14l5,gm459...   \n",
       "4  0.821667  ermn,creb5,c030029h02rik,tsc22d4,cpm,d7ertd443...   \n",
       "5  0.786952  prokr2,npy,6530411m01rik,npas1,catip,slc32a1,g...   \n",
       "6  0.821310  ac034116.3,phka1,ppara,nat8f3,ccdc141,gm35552,...   \n",
       "7  0.955714  crip1,cp,cavin1,myh11,cfh,pf4,gm47719,mpzl2,s1...   \n",
       "\n",
       "                                    GO Module  #pc_genes  \\\n",
       "0                                angiogenesis       59.0   \n",
       "1     cellular response to lipopolysaccharide       32.0   \n",
       "2                         response to hormone       61.0   \n",
       "3                     ensheathment of neurons       76.0   \n",
       "4                     ensheathment of neurons       92.0   \n",
       "5                 glutamate catabolic process       31.0   \n",
       "6  regulation of smoothened signaling pathway       27.0   \n",
       "7         positive regulation of MAPK cascade       79.0   \n",
       "\n",
       "                                             GO Cell  \n",
       "0                       collagen fibril organization  \n",
       "1                  extracellular matrix organization  \n",
       "2                        skeletal system development  \n",
       "3                                        myelination  \n",
       "4                            ensheathment of neurons  \n",
       "5  cerebral cortex GABAergic interneuron differen...  \n",
       "6  regulation of generation of precursor metaboli...  \n",
       "7                                 tissue development  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_st.uns['instant_biclustering']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "98d5bb5d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>method</th>\n",
       "      <th>genes</th>\n",
       "      <th>uIDs</th>\n",
       "      <th>#cells</th>\n",
       "      <th>combined</th>\n",
       "      <th>pre cell expansion</th>\n",
       "      <th>post cell expansion</th>\n",
       "      <th>sprawl average</th>\n",
       "      <th>sprawl score</th>\n",
       "      <th>tangram</th>\n",
       "      <th>tangram_corr</th>\n",
       "      <th>baseline</th>\n",
       "      <th>shap genes</th>\n",
       "      <th>GO Module</th>\n",
       "      <th>#pc_genes</th>\n",
       "      <th>GO Cell</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Peripheral</td>\n",
       "      <td>slc13a4,aldh1a2,spp1,fmod,dcn,igf2,acta2</td>\n",
       "      <td>8215,140712,126630,136597,155150,96880,57662,1...</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>1627.962243</td>\n",
       "      <td>1627.962243</td>\n",
       "      <td>0.21273664077562526</td>\n",
       "      <td>1627.9622432729848</td>\n",
       "      <td>0.848526</td>\n",
       "      <td>0.843269</td>\n",
       "      <td>0.870192</td>\n",
       "      <td>ogn,bmp6,h2-ab1,cfh,myoc,slc13a3,cp,zic2,fam18...</td>\n",
       "      <td>regulation of vascular endothelial cell prolif...</td>\n",
       "      <td>65.0</td>\n",
       "      <td>skeletal system development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Radial</td>\n",
       "      <td>cldn5,adgrl4,pecam1,acvrl1,ly6a,fn1,pglyrp1,so...</td>\n",
       "      <td>76365,141457,80575,61374,115711,108543,94090,4...</td>\n",
       "      <td>181</td>\n",
       "      <td>0</td>\n",
       "      <td>4501.068575</td>\n",
       "      <td>4501.068575</td>\n",
       "      <td>0.34304887377815557</td>\n",
       "      <td>4501.068575461013</td>\n",
       "      <td>0.904369</td>\n",
       "      <td>0.901622</td>\n",
       "      <td>0.858919</td>\n",
       "      <td>slc6a20a,4930523c07rik,gjb2,asgr1,itih2,gm4267...</td>\n",
       "      <td>angiogenesis</td>\n",
       "      <td>56.0</td>\n",
       "      <td>renal system development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Radial</td>\n",
       "      <td>cd53,trem2,cd300c2,siglech,cd68,ikzf1,spi1,laptm5</td>\n",
       "      <td>19695,94774,42871,136552,40598,112643,42585,41...</td>\n",
       "      <td>89</td>\n",
       "      <td>0</td>\n",
       "      <td>4192.935669</td>\n",
       "      <td>4192.935669</td>\n",
       "      <td>0.46629213483146065</td>\n",
       "      <td>4192.935668805927</td>\n",
       "      <td>0.885425</td>\n",
       "      <td>0.887124</td>\n",
       "      <td>0.823595</td>\n",
       "      <td>crybb1,ltc4s,c1qc,ajuba,tyrobp,lyl1,il21r,fcgr...</td>\n",
       "      <td>regulation of immune system process</td>\n",
       "      <td>32.0</td>\n",
       "      <td>complement activation, classical pathway</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Radial</td>\n",
       "      <td>aldh1a2,col1a1,fmod,spp1,gjb2,cyp1b1,slc13a4,d...</td>\n",
       "      <td>155150,137037,14188,140712,120166,143128,8932,...</td>\n",
       "      <td>79</td>\n",
       "      <td>0</td>\n",
       "      <td>2197.957627</td>\n",
       "      <td>2197.957627</td>\n",
       "      <td>0.3308917018284107</td>\n",
       "      <td>2197.95762713589</td>\n",
       "      <td>0.915458</td>\n",
       "      <td>0.907750</td>\n",
       "      <td>0.919667</td>\n",
       "      <td>cp,tbx15,prelp,asgr1,fgl2,lbp,foxc2,slc26a7,wn...</td>\n",
       "      <td>response to nutrient</td>\n",
       "      <td>60.0</td>\n",
       "      <td>skeletal system development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Radial</td>\n",
       "      <td>ano1,carmn,cspg4</td>\n",
       "      <td>10612,57455,64726,46017,107900,149206,61426,40...</td>\n",
       "      <td>57</td>\n",
       "      <td>0</td>\n",
       "      <td>2145.904307</td>\n",
       "      <td>2145.904307</td>\n",
       "      <td>0.5977076023391813</td>\n",
       "      <td>2145.904306614343</td>\n",
       "      <td>0.825833</td>\n",
       "      <td>0.816136</td>\n",
       "      <td>0.927045</td>\n",
       "      <td>gata2,a2m,pcolce,osr1,b230323a14rik,hspg2,slc2...</td>\n",
       "      <td>gliogenesis</td>\n",
       "      <td>56.0</td>\n",
       "      <td>renal system development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Radial</td>\n",
       "      <td>gpr17,cspg4,sema3d,pdgfra</td>\n",
       "      <td>94389,11221,104206,1377,73950,85223,130273,103...</td>\n",
       "      <td>104</td>\n",
       "      <td>0</td>\n",
       "      <td>1938.946652</td>\n",
       "      <td>1957.429210</td>\n",
       "      <td>0.5515961538461538</td>\n",
       "      <td>1957.4292098625492</td>\n",
       "      <td>0.781024</td>\n",
       "      <td>0.782381</td>\n",
       "      <td>0.751786</td>\n",
       "      <td>litaf,tmem98,arrdc2,pla2g16,gm33594,gatm,prr5l...</td>\n",
       "      <td>platelet-derived growth factor receptor signal...</td>\n",
       "      <td>91.0</td>\n",
       "      <td>myelination</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Punctate</td>\n",
       "      <td>cldn5,adgrl4,pecam1,acvrl1,ly6a,pglyrp1,fn1,kd...</td>\n",
       "      <td>91165,10170,57455,53774,108270,54266,17139,420...</td>\n",
       "      <td>204</td>\n",
       "      <td>0</td>\n",
       "      <td>6981.428102</td>\n",
       "      <td>6981.428102</td>\n",
       "      <td>0.40345588235294116</td>\n",
       "      <td>6981.428102117615</td>\n",
       "      <td>0.930476</td>\n",
       "      <td>0.927756</td>\n",
       "      <td>0.890280</td>\n",
       "      <td>col1a2,fmod,lum,itih2,igf2,pf4,gjb2,fam180a,co...</td>\n",
       "      <td>angiogenesis</td>\n",
       "      <td>54.0</td>\n",
       "      <td>renal system development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Punctate</td>\n",
       "      <td>cd300c2,cd53,trem2,ikzf1,siglech,spi1,cd68,laptm5</td>\n",
       "      <td>19695,63439,84448,94774,88506,112643,10693,128...</td>\n",
       "      <td>99</td>\n",
       "      <td>0</td>\n",
       "      <td>6559.742852</td>\n",
       "      <td>6559.742852</td>\n",
       "      <td>0.5703484848484849</td>\n",
       "      <td>6559.742851723009</td>\n",
       "      <td>0.870868</td>\n",
       "      <td>0.872395</td>\n",
       "      <td>0.810711</td>\n",
       "      <td>tyrobp,ang,cdca7,vsir,p3h4,pllp,crybb1,gm44532...</td>\n",
       "      <td>regulation of immune system process</td>\n",
       "      <td>58.0</td>\n",
       "      <td>ensheathment of neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Punctate</td>\n",
       "      <td>cspg4,gpr17,ano1,carmn,sema3d,pdgfra</td>\n",
       "      <td>46017,42508,61426,57455,136504,64726,103198,94...</td>\n",
       "      <td>193</td>\n",
       "      <td>0</td>\n",
       "      <td>2828.799095</td>\n",
       "      <td>2828.799095</td>\n",
       "      <td>0.38832469775474954</td>\n",
       "      <td>2828.79909467086</td>\n",
       "      <td>0.819845</td>\n",
       "      <td>0.820081</td>\n",
       "      <td>0.838354</td>\n",
       "      <td>cldn14,gm10687,2700046a07rik,serpinb1a,gm44866...</td>\n",
       "      <td>platelet-derived growth factor receptor signal...</td>\n",
       "      <td>84.0</td>\n",
       "      <td>ensheathment of neurons</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Punctate</td>\n",
       "      <td>fmod,col1a1,spp1,aldh1a2,dcn,gjb2,cyp1b1,igf2,...</td>\n",
       "      <td>155150,137037,143128,26240,93733,40698,140712,...</td>\n",
       "      <td>88</td>\n",
       "      <td>0</td>\n",
       "      <td>2340.250294</td>\n",
       "      <td>2340.250294</td>\n",
       "      <td>0.3084454545454546</td>\n",
       "      <td>2340.2502944974276</td>\n",
       "      <td>0.904673</td>\n",
       "      <td>0.886405</td>\n",
       "      <td>0.911993</td>\n",
       "      <td>itih2,slc6a12,aebp1,serpind1,slc6a13,emp3,lbp,...</td>\n",
       "      <td>response to nutrient</td>\n",
       "      <td>51.0</td>\n",
       "      <td>nephron development</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Punctate</td>\n",
       "      <td>sox10,opalin,gjc3,zfp536,sema6a</td>\n",
       "      <td>27905,103084,148537,67016,126980,18603,48454,7...</td>\n",
       "      <td>307</td>\n",
       "      <td>0</td>\n",
       "      <td>2175.577612</td>\n",
       "      <td>2175.577612</td>\n",
       "      <td>0.6195908794788274</td>\n",
       "      <td>2175.577611765847</td>\n",
       "      <td>0.914194</td>\n",
       "      <td>0.913525</td>\n",
       "      <td>0.904418</td>\n",
       "      <td>sec14l5,gm4593,gm42756,gm44866,gm48508,cyp2j12...</td>\n",
       "      <td>ensheathment of neurons</td>\n",
       "      <td>77.0</td>\n",
       "      <td>myelination</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Punctate</td>\n",
       "      <td>cort,sst,pthlh,kcnmb2,rbp4,gad2,crh,rab3b,chod...</td>\n",
       "      <td>144670,53723,159779,129930,129048,143684,64454...</td>\n",
       "      <td>126</td>\n",
       "      <td>0</td>\n",
       "      <td>2029.809574</td>\n",
       "      <td>2026.653664</td>\n",
       "      <td>0.2005414462081129</td>\n",
       "      <td>2026.653664390531</td>\n",
       "      <td>0.831800</td>\n",
       "      <td>0.821354</td>\n",
       "      <td>0.835369</td>\n",
       "      <td>lhx6,dlx2,dlx6os1,dlx1as,lypd6,rpp25,cox6a2,nd...</td>\n",
       "      <td>hormone-mediated signaling pathway</td>\n",
       "      <td>22.0</td>\n",
       "      <td>cerebral cortex GABAergic interneuron differen...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Central</td>\n",
       "      <td>cldn5,pglyrp1,ly6a,pecam1,acvrl1,adgrl4,emcn,f...</td>\n",
       "      <td>35357,80575,57455,9201,33173,108270,2908,11771...</td>\n",
       "      <td>150</td>\n",
       "      <td>0</td>\n",
       "      <td>5457.167341</td>\n",
       "      <td>5457.167341</td>\n",
       "      <td>0.21660920077508525</td>\n",
       "      <td>5457.167341076012</td>\n",
       "      <td>0.907000</td>\n",
       "      <td>0.905333</td>\n",
       "      <td>0.865333</td>\n",
       "      <td>myoc,usp18,fmod,slc6a12,hbb-bt,slc6a13,slc13a4...</td>\n",
       "      <td>angiogenesis</td>\n",
       "      <td>72.0</td>\n",
       "      <td>collagen fibril organization</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Central</td>\n",
       "      <td>trem2,cd300c2,cd53,siglech,ikzf1,cd68,spi1,laptm5</td>\n",
       "      <td>128059,10693,84448,88506,107354,62942,144249,1...</td>\n",
       "      <td>93</td>\n",
       "      <td>0</td>\n",
       "      <td>4816.045183</td>\n",
       "      <td>4816.045183</td>\n",
       "      <td>0.30136336685376397</td>\n",
       "      <td>4816.045182517613</td>\n",
       "      <td>0.883041</td>\n",
       "      <td>0.884064</td>\n",
       "      <td>0.808363</td>\n",
       "      <td>cdca7,tyrobp,gpr84,icosl,ltc4s,gm42674,agpat2,...</td>\n",
       "      <td>regulation of immune system process</td>\n",
       "      <td>20.0</td>\n",
       "      <td>positive regulation of activated T cell prolif...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Central</td>\n",
       "      <td>cspg4,gpr17,carmn,ano1,sema3d,pdgfra</td>\n",
       "      <td>6387,3852,85030,80885,40275,54807,15304,103907...</td>\n",
       "      <td>136</td>\n",
       "      <td>0</td>\n",
       "      <td>2338.515502</td>\n",
       "      <td>2338.515502</td>\n",
       "      <td>0.2284871047503147</td>\n",
       "      <td>2338.515502237935</td>\n",
       "      <td>0.804788</td>\n",
       "      <td>0.805847</td>\n",
       "      <td>0.829233</td>\n",
       "      <td>2700046a07rik,hspg2,gm44866,serpinb1a,apod,gm1...</td>\n",
       "      <td>platelet-derived growth factor receptor signal...</td>\n",
       "      <td>74.0</td>\n",
       "      <td>myelination</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Central</td>\n",
       "      <td>opalin,sox10,gjc3,zfp536,sema6a,dpy19l1,prox1</td>\n",
       "      <td>103084,126980,2446,148537,66207,150209,6857,86...</td>\n",
       "      <td>193</td>\n",
       "      <td>0</td>\n",
       "      <td>1999.216194</td>\n",
       "      <td>1999.216194</td>\n",
       "      <td>0.34536512529970403</td>\n",
       "      <td>1999.2161936838038</td>\n",
       "      <td>0.866356</td>\n",
       "      <td>0.860958</td>\n",
       "      <td>0.853131</td>\n",
       "      <td>cyp2j12,gm4593,galnt6,sec14l5,myrf,a330049n07r...</td>\n",
       "      <td>ameboidal-type cell migration</td>\n",
       "      <td>82.0</td>\n",
       "      <td>ensheathment of neurons</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        method                                              genes  \\\n",
       "0   Peripheral           slc13a4,aldh1a2,spp1,fmod,dcn,igf2,acta2   \n",
       "1       Radial  cldn5,adgrl4,pecam1,acvrl1,ly6a,fn1,pglyrp1,so...   \n",
       "2       Radial  cd53,trem2,cd300c2,siglech,cd68,ikzf1,spi1,laptm5   \n",
       "3       Radial  aldh1a2,col1a1,fmod,spp1,gjb2,cyp1b1,slc13a4,d...   \n",
       "4       Radial                                   ano1,carmn,cspg4   \n",
       "5       Radial                          gpr17,cspg4,sema3d,pdgfra   \n",
       "6     Punctate  cldn5,adgrl4,pecam1,acvrl1,ly6a,pglyrp1,fn1,kd...   \n",
       "7     Punctate  cd300c2,cd53,trem2,ikzf1,siglech,spi1,cd68,laptm5   \n",
       "8     Punctate               cspg4,gpr17,ano1,carmn,sema3d,pdgfra   \n",
       "9     Punctate  fmod,col1a1,spp1,aldh1a2,dcn,gjb2,cyp1b1,igf2,...   \n",
       "10    Punctate                    sox10,opalin,gjc3,zfp536,sema6a   \n",
       "11    Punctate  cort,sst,pthlh,kcnmb2,rbp4,gad2,crh,rab3b,chod...   \n",
       "12     Central  cldn5,pglyrp1,ly6a,pecam1,acvrl1,adgrl4,emcn,f...   \n",
       "13     Central  trem2,cd300c2,cd53,siglech,ikzf1,cd68,spi1,laptm5   \n",
       "14     Central               cspg4,gpr17,carmn,ano1,sema3d,pdgfra   \n",
       "15     Central      opalin,sox10,gjc3,zfp536,sema6a,dpy19l1,prox1   \n",
       "\n",
       "                                                 uIDs  #cells  combined  \\\n",
       "0   8215,140712,126630,136597,155150,96880,57662,1...      61         0   \n",
       "1   76365,141457,80575,61374,115711,108543,94090,4...     181         0   \n",
       "2   19695,94774,42871,136552,40598,112643,42585,41...      89         0   \n",
       "3   155150,137037,14188,140712,120166,143128,8932,...      79         0   \n",
       "4   10612,57455,64726,46017,107900,149206,61426,40...      57         0   \n",
       "5   94389,11221,104206,1377,73950,85223,130273,103...     104         0   \n",
       "6   91165,10170,57455,53774,108270,54266,17139,420...     204         0   \n",
       "7   19695,63439,84448,94774,88506,112643,10693,128...      99         0   \n",
       "8   46017,42508,61426,57455,136504,64726,103198,94...     193         0   \n",
       "9   155150,137037,143128,26240,93733,40698,140712,...      88         0   \n",
       "10  27905,103084,148537,67016,126980,18603,48454,7...     307         0   \n",
       "11  144670,53723,159779,129930,129048,143684,64454...     126         0   \n",
       "12  35357,80575,57455,9201,33173,108270,2908,11771...     150         0   \n",
       "13  128059,10693,84448,88506,107354,62942,144249,1...      93         0   \n",
       "14  6387,3852,85030,80885,40275,54807,15304,103907...     136         0   \n",
       "15  103084,126980,2446,148537,66207,150209,6857,86...     193         0   \n",
       "\n",
       "    pre cell expansion  post cell expansion       sprawl average  \\\n",
       "0          1627.962243          1627.962243  0.21273664077562526   \n",
       "1          4501.068575          4501.068575  0.34304887377815557   \n",
       "2          4192.935669          4192.935669  0.46629213483146065   \n",
       "3          2197.957627          2197.957627   0.3308917018284107   \n",
       "4          2145.904307          2145.904307   0.5977076023391813   \n",
       "5          1938.946652          1957.429210   0.5515961538461538   \n",
       "6          6981.428102          6981.428102  0.40345588235294116   \n",
       "7          6559.742852          6559.742852   0.5703484848484849   \n",
       "8          2828.799095          2828.799095  0.38832469775474954   \n",
       "9          2340.250294          2340.250294   0.3084454545454546   \n",
       "10         2175.577612          2175.577612   0.6195908794788274   \n",
       "11         2029.809574          2026.653664   0.2005414462081129   \n",
       "12         5457.167341          5457.167341  0.21660920077508525   \n",
       "13         4816.045183          4816.045183  0.30136336685376397   \n",
       "14         2338.515502          2338.515502   0.2284871047503147   \n",
       "15         1999.216194          1999.216194  0.34536512529970403   \n",
       "\n",
       "          sprawl score   tangram  tangram_corr  baseline  \\\n",
       "0   1627.9622432729848  0.848526      0.843269  0.870192   \n",
       "1    4501.068575461013  0.904369      0.901622  0.858919   \n",
       "2    4192.935668805927  0.885425      0.887124  0.823595   \n",
       "3     2197.95762713589  0.915458      0.907750  0.919667   \n",
       "4    2145.904306614343  0.825833      0.816136  0.927045   \n",
       "5   1957.4292098625492  0.781024      0.782381  0.751786   \n",
       "6    6981.428102117615  0.930476      0.927756  0.890280   \n",
       "7    6559.742851723009  0.870868      0.872395  0.810711   \n",
       "8     2828.79909467086  0.819845      0.820081  0.838354   \n",
       "9   2340.2502944974276  0.904673      0.886405  0.911993   \n",
       "10   2175.577611765847  0.914194      0.913525  0.904418   \n",
       "11   2026.653664390531  0.831800      0.821354  0.835369   \n",
       "12   5457.167341076012  0.907000      0.905333  0.865333   \n",
       "13   4816.045182517613  0.883041      0.884064  0.808363   \n",
       "14   2338.515502237935  0.804788      0.805847  0.829233   \n",
       "15  1999.2161936838038  0.866356      0.860958  0.853131   \n",
       "\n",
       "                                           shap genes  \\\n",
       "0   ogn,bmp6,h2-ab1,cfh,myoc,slc13a3,cp,zic2,fam18...   \n",
       "1   slc6a20a,4930523c07rik,gjb2,asgr1,itih2,gm4267...   \n",
       "2   crybb1,ltc4s,c1qc,ajuba,tyrobp,lyl1,il21r,fcgr...   \n",
       "3   cp,tbx15,prelp,asgr1,fgl2,lbp,foxc2,slc26a7,wn...   \n",
       "4   gata2,a2m,pcolce,osr1,b230323a14rik,hspg2,slc2...   \n",
       "5   litaf,tmem98,arrdc2,pla2g16,gm33594,gatm,prr5l...   \n",
       "6   col1a2,fmod,lum,itih2,igf2,pf4,gjb2,fam180a,co...   \n",
       "7   tyrobp,ang,cdca7,vsir,p3h4,pllp,crybb1,gm44532...   \n",
       "8   cldn14,gm10687,2700046a07rik,serpinb1a,gm44866...   \n",
       "9   itih2,slc6a12,aebp1,serpind1,slc6a13,emp3,lbp,...   \n",
       "10  sec14l5,gm4593,gm42756,gm44866,gm48508,cyp2j12...   \n",
       "11  lhx6,dlx2,dlx6os1,dlx1as,lypd6,rpp25,cox6a2,nd...   \n",
       "12  myoc,usp18,fmod,slc6a12,hbb-bt,slc6a13,slc13a4...   \n",
       "13  cdca7,tyrobp,gpr84,icosl,ltc4s,gm42674,agpat2,...   \n",
       "14  2700046a07rik,hspg2,gm44866,serpinb1a,apod,gm1...   \n",
       "15  cyp2j12,gm4593,galnt6,sec14l5,myrf,a330049n07r...   \n",
       "\n",
       "                                            GO Module  #pc_genes  \\\n",
       "0   regulation of vascular endothelial cell prolif...       65.0   \n",
       "1                                        angiogenesis       56.0   \n",
       "2                 regulation of immune system process       32.0   \n",
       "3                                response to nutrient       60.0   \n",
       "4                                         gliogenesis       56.0   \n",
       "5   platelet-derived growth factor receptor signal...       91.0   \n",
       "6                                        angiogenesis       54.0   \n",
       "7                 regulation of immune system process       58.0   \n",
       "8   platelet-derived growth factor receptor signal...       84.0   \n",
       "9                                response to nutrient       51.0   \n",
       "10                            ensheathment of neurons       77.0   \n",
       "11                 hormone-mediated signaling pathway       22.0   \n",
       "12                                       angiogenesis       72.0   \n",
       "13                regulation of immune system process       20.0   \n",
       "14  platelet-derived growth factor receptor signal...       74.0   \n",
       "15                      ameboidal-type cell migration       82.0   \n",
       "\n",
       "                                              GO Cell  \n",
       "0                         skeletal system development  \n",
       "1                            renal system development  \n",
       "2            complement activation, classical pathway  \n",
       "3                         skeletal system development  \n",
       "4                            renal system development  \n",
       "5                                         myelination  \n",
       "6                            renal system development  \n",
       "7                             ensheathment of neurons  \n",
       "8                             ensheathment of neurons  \n",
       "9                                 nephron development  \n",
       "10                                        myelination  \n",
       "11  cerebral cortex GABAergic interneuron differen...  \n",
       "12                       collagen fibril organization  \n",
       "13  positive regulation of activated T cell prolif...  \n",
       "14                                        myelination  \n",
       "15                            ensheathment of neurons  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_st.uns['sprawl_biclustering']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "133525a4",
   "metadata": {},
   "source": [
    "---\n",
    "\n",
    "#### Module Characterization Outputs\n",
    "\n",
    "Outputs from the module characterization of the modules using GO enrichment tests. Outputs are presented as `DataFrames` contained within a dictionary. `instant_biclustering_geo_{x}` and `sprawl_biclustering_geo_{x}` are dictionaries with keys as the module number (str) and the DataFrames are keys. The DataFrames are the standard `csv` that are outputed by PantherDB overrepresentation test. `x` is either `cell` or `module` depending on the cell or genes characterization respectively.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "7aa185f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>number_in_list</th>\n",
       "      <th>fold_enrichment</th>\n",
       "      <th>fdr</th>\n",
       "      <th>expected</th>\n",
       "      <th>number_in_reference</th>\n",
       "      <th>pValue</th>\n",
       "      <th>term</th>\n",
       "      <th>id</th>\n",
       "      <th>dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>6.068111</td>\n",
       "      <td>0.018886</td>\n",
       "      <td>1.318367</td>\n",
       "      <td>17</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>angiogenesis</td>\n",
       "      <td>GO:0001525</td>\n",
       "      <td>BP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "      <td>4.912281</td>\n",
       "      <td>0.066385</td>\n",
       "      <td>1.628571</td>\n",
       "      <td>21</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>blood vessel morphogenesis</td>\n",
       "      <td>GO:0048514</td>\n",
       "      <td>BP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8</td>\n",
       "      <td>3.967611</td>\n",
       "      <td>0.272274</td>\n",
       "      <td>2.016327</td>\n",
       "      <td>26</td>\n",
       "      <td>0.000189</td>\n",
       "      <td>blood vessel development</td>\n",
       "      <td>GO:0001568</td>\n",
       "      <td>BP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9</td>\n",
       "      <td>3.413313</td>\n",
       "      <td>0.251671</td>\n",
       "      <td>2.636735</td>\n",
       "      <td>34</td>\n",
       "      <td>0.000232</td>\n",
       "      <td>tube morphogenesis</td>\n",
       "      <td>GO:0035239</td>\n",
       "      <td>BP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>8</td>\n",
       "      <td>3.820663</td>\n",
       "      <td>0.221991</td>\n",
       "      <td>2.093878</td>\n",
       "      <td>27</td>\n",
       "      <td>0.000256</td>\n",
       "      <td>vasculature development</td>\n",
       "      <td>GO:0001944</td>\n",
       "      <td>BP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>2</td>\n",
       "      <td>5.157895</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.387755</td>\n",
       "      <td>5</td>\n",
       "      <td>0.049595</td>\n",
       "      <td>transforming growth factor beta receptor super...</td>\n",
       "      <td>GO:0141091</td>\n",
       "      <td>BP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>2</td>\n",
       "      <td>5.157895</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.387755</td>\n",
       "      <td>5</td>\n",
       "      <td>0.049595</td>\n",
       "      <td>transforming growth factor beta receptor signa...</td>\n",
       "      <td>GO:0007179</td>\n",
       "      <td>BP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>2</td>\n",
       "      <td>5.157895</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.387755</td>\n",
       "      <td>5</td>\n",
       "      <td>0.049595</td>\n",
       "      <td>cell surface receptor protein serine/threonine...</td>\n",
       "      <td>GO:0007178</td>\n",
       "      <td>BP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80</th>\n",
       "      <td>2</td>\n",
       "      <td>5.157895</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.387755</td>\n",
       "      <td>5</td>\n",
       "      <td>0.049595</td>\n",
       "      <td>positive regulation of vasculature development</td>\n",
       "      <td>GO:1904018</td>\n",
       "      <td>BP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2</td>\n",
       "      <td>5.157895</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.387755</td>\n",
       "      <td>5</td>\n",
       "      <td>0.049595</td>\n",
       "      <td>carbohydrate binding</td>\n",
       "      <td>GO:0030246</td>\n",
       "      <td>MF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>97 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    number_in_list  fold_enrichment       fdr  expected  number_in_reference  \\\n",
       "0                8         6.068111  0.018886  1.318367                   17   \n",
       "1                8         4.912281  0.066385  1.628571                   21   \n",
       "2                8         3.967611  0.272274  2.016327                   26   \n",
       "3                9         3.413313  0.251671  2.636735                   34   \n",
       "4                8         3.820663  0.221991  2.093878                   27   \n",
       "..             ...              ...       ...       ...                  ...   \n",
       "79               2         5.157895  1.000000  0.387755                    5   \n",
       "81               2         5.157895  1.000000  0.387755                    5   \n",
       "82               2         5.157895  1.000000  0.387755                    5   \n",
       "80               2         5.157895  1.000000  0.387755                    5   \n",
       "5                2         5.157895  1.000000  0.387755                    5   \n",
       "\n",
       "      pValue                                               term          id  \\\n",
       "0   0.000004                                       angiogenesis  GO:0001525   \n",
       "1   0.000031                         blood vessel morphogenesis  GO:0048514   \n",
       "2   0.000189                           blood vessel development  GO:0001568   \n",
       "3   0.000232                                 tube morphogenesis  GO:0035239   \n",
       "4   0.000256                            vasculature development  GO:0001944   \n",
       "..       ...                                                ...         ...   \n",
       "79  0.049595  transforming growth factor beta receptor super...  GO:0141091   \n",
       "81  0.049595  transforming growth factor beta receptor signa...  GO:0007179   \n",
       "82  0.049595  cell surface receptor protein serine/threonine...  GO:0007178   \n",
       "80  0.049595     positive regulation of vasculature development  GO:1904018   \n",
       "5   0.049595                               carbohydrate binding  GO:0030246   \n",
       "\n",
       "   dataset  \n",
       "0       BP  \n",
       "1       BP  \n",
       "2       BP  \n",
       "3       BP  \n",
       "4       BP  \n",
       "..     ...  \n",
       "79      BP  \n",
       "81      BP  \n",
       "82      BP  \n",
       "80      BP  \n",
       "5       MF  \n",
       "\n",
       "[97 rows x 9 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_st.uns['instant_biclustering_geo_module']['0']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:.conda-cellSP_env]",
   "language": "python",
   "name": "conda-env-.conda-cellSP_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
